pkg/scope/health/health.go

```diff
@@ -0,0 +1,148 @@
+func UpdateHealthStatus(ctx context.Context, client client.Client, healthScope *v1alpha2.HealthScope) error {
+	timeout := defaultTimeout
+	if healthScope.Spec.ProbeTimeout != nil {
+		timeout = time.Duration(*healthScope.Spec.ProbeTimeout) * time.Second
+	}
+
+	resourceRefs := []runtimev1alpha1.TypedReference{}
+	for _, workloadRef := range healthScope.Status.WorkloadReferences {
+		// Get workload object.
+		workloadObject := unstructured.Unstructured{}
+		workloadObject.SetAPIVersion(workloadRef.APIVersion)
+		workloadObject.SetKind(workloadRef.Kind)
+		workloadObjectRef := types.NamespacedName{Namespace: healthScope.GetNamespace(), Name: workloadRef.Name}
+		if err := client.Get(ctx, workloadObjectRef, &workloadObject); err != nil {
+			return errors.Wrapf(err, errNoWorkload, workloadRef.Name)
+		}
+
+		// TODO(asouza): Need to find a way to make sure that this is compatible with the scope structure.
+		if value, err := fieldpath.Pave(workloadObject.UnstructuredContent()).GetValue("status.resources"); err == nil {
+			refs := value.([]interface{})
+			for _, item := range refs {
+				ref := item.(map[string]interface{})
+				resourceRef := runtimev1alpha1.TypedReference{
+					APIVersion: fmt.Sprintf("%v", ref["apiVersion"]),
+					Kind:       fmt.Sprintf("%v", ref["kind"]),
+					Name:       fmt.Sprintf("%v", ref["name"]),
+				}
+
+				resourceRefs = append(resourceRefs, resourceRef)
+			}
+		} else {
+			return errors.Wrapf(err, errNoWorkloadResources, workloadRef.Name)
+		}
+	}
+
+	statusc := resourcesHealthStatus(ctx, client, healthScope.Namespace, resourceRefs, timeout)
+	status := true
+	for r := range statusc {
+		status = status && r
+	}
+
+	health := "unhealthy"
+	if status {
+		health = "healthy"
+	}
+
+	healthScope.Status.Health = health
+	return nil
+}
+
+func resourcesHealthStatus(ctx context.Context, client client.Client, namespace string, refs []runtimev1alpha1.TypedReference, timeout time.Duration) <-chan bool {
+	status := make(chan bool)
+	var wg sync.WaitGroup
+	wg.Add(len(refs))
+	for _, ref := range refs {
+		go func(resourceRef runtimev1alpha1.TypedReference) {
+			defer wg.Done()
+			err := resourceHealthStatus(ctx, client, namespace, resourceRef, timeout)
+			status <- (err == nil)
+			if err != nil {
+				fmt.Printf("Unhealthy resource %q: %v\n", resourceRef.Name, err)
+			}
+		}(ref)
+	}
+	go func() {
+		wg.Wait()
+		close(status)
+	}()
+
+	return status
+}
+
+func resourceHealthStatus(ctx context.Context, client client.Client, namespace string, ref runtimev1alpha1.TypedReference, _ time.Duration) error {
+	// TODO(artursouza): figure out how to apply timeout.
+	if (ref.APIVersion == "apps/v1") && (ref.Kind == "Deployment") {
+		return deploymentHealthStatus(ctx, client, namespace, ref)
+	}
+
+	// TODO(artursouza): add other health checks.
+	// Generic health check by validating if the resource exists.
+	object := unstructured.Unstructured{}
```

Comment:
what about the health probe?

Reply:
We discussed about it last week. We must extend the spec to have health check specific per workload type. For unknown types, we thought about adding a trait to specify which health check to perform, like in a Database, for example.

This is another work item (IMO).

Reply:
We can at least do some duck typing to check on status